NewGameChooser - Cancel Map Zip Parsing When The Game Starts (clear is called)

DanVanAtta opened the pull request:
Allows games to start without having to wait for background map parsing to complete, that data is cleared anyways when a game starts. We do this by making the clear operation non-blocking, we now set a flag to notify refresh operations that they should ignore their results when done computing. Secondly, we will also actively cancel refresh operations to avoid unnecessary resource usage.

For context: map data is loaded in the background when tripleA is launched or when a user leaves a game. This data is used to pre-compute the game selection window, letting a user choose a game quickly.  To save memory, we remove this preloaded background data while a user is playing a game. By restarting a local game rapidly, you can notice the delay of loading map data (I had 5 seconds with 11 maps on a powerful 6 core CPU with rotating rotating hard disks)

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/triplea-game/triplea/74)
<!-- Reviewable:end -->

DanVanAtta commented on src/games/strategy/engine/framework/ui/NewGameChooser.java:
```diff
@@ -238,20 +245,29 @@ public synchronized static NewGameChooserModel getNewGameChooserModel() {
+    NewGameChooserModel model = new NewGameChooserModel(cacheClearedMessenger);
+      // Cache might not be populated if another thread has cleared
+      // the cache while we were loading.
+    if (mapParsingInProgress) {
```
This if statement is really critical to the change. It allows us to remove the synchronization from the clear method. The clear method will instead set this flag and the clear method can then be unblocked to do other stuff. Meanwhile, if clear has been called while we are still building a NewGameChooserModel, then we'll see the flag has been changed and we'll ignore the value of "model" altogether. This is desired, we want that model object to be GC'd so it doesn't consume memory while we are playing.

ron-murhammer replied:
@DanVanAtta I'm not sure that this works. So the scenario I'm thinking of is as follows:

1. Thread A calls refresh and gets into the if statement but hasn't executed 's_cachedGameModel = model' yet
2. Thread B calls clear and sets s_cachedGameModel to null
3. Thread A resumes and then overrides the clear by setting s_cachedGameModel to the new model

DanVanAtta replied:
A lock at lines 255 and 264 around "mapParsingInProgress" should fix that. The lock is being held for just in-memory operations, so it'll be released quickly and won't delay game launch.

There are some scenarios that are impossible since both threads would have to be driven by simultaneous user interaction. I can't recall which scenarios exactly that is the case, but doesn't seem to matter for this specific case.

veqryn commented:
Looks promising, but I'll do a full review of this after the release.

ron-murhammer commented:
@DanVanAtta Can you rebase this one? Looks good at a high level but will need to review a little closer.

DanVanAtta commented:
Rebased, but lack of #151 forced a line endings fix commit.

DanVanAtta commented:
@ron-murhammer , I updated the PR to add in the lock check to prevent the problem you pointed out. I also retested game launching and selection, things look good.

ron-murhammer commented:
@DanVanAtta I still don't think that solves all the issues. Can't the following still occur:
1. refresh is called but doesn't execute any lines yet
2. clear is called and runs all the way through
3. refresh picks back up and runs all the way through overriding the clear

DanVanAtta commented:
That cannot occur. The first thing clear does is set `mapParsingInProgress = false;`

When refresh picks back up, this code will not override the clear:
```
+    synchronized( mapParsingInProgressLock ) {
+      if (mapParsingInProgress) {
+        s_cachedGameModel = model;
+      }
+    }
```

What we've done in this PR is to allow clear to terminate before refresh. To accommodate for the now potentially asynchronous ordering - we have a flag `mapParsingInProgress` so the `clear` method can notify the `refresh` operation that it should throw away its results.

We do still have a lock to avoid a race condition after map parsing has finished (thanks for spotting this @ron-murhammer). The lock moving is really the heart of the change. Instead of grabbing it before we start map parsing, which can take seconds, we instead grab it when map parsing is done. So we still have a lock to solve the sequencing problem between refresh and clear, but instead now we hold it for only in-memory operations rather than operations that involve many hard disk fetches.

ron-murhammer commented:
@DanVanAtta Ok I think you're right. But can't calling clear 2 times in a row cause a NPE on cacheClearedMessenger?

I think the changes have become overly complex. I think we should step back and define exactly what we are trying to accomplish with this PR. Would it be accurate to say the goal is to avoid the several second pause due to map parsing done in refresh when clear is called? So essentially we want to abort the refresh ASAP when clear is called?

If that is that case then could we just simplify the clear to just set the cancel on the messenger then wait for refresh to finish. So essentially all of clear stays synchronized except the first line would set cancel on the messenger.

DanVanAtta commented:
Kindly restart the game as fast you can and notice the delay. You'll notice a delay in the start of the game. This delay is because clear is blocking until refresh completes. We know clear just throws away the data refresh generates, so blocking until you can throw away data, causing a delay to the start of the game is a bug.

Perhaps we avoided other bugs by having these locks, but this one exists. It's also quite clear the use case of launch a game, click choose game, start game was what was considered when the background loading feature was added. It's not clear if it was well considered what the implications would be for starting a game when the map parsing time was 5 or 10 seconds or even slower.



"So essentially we want to abort the refresh ASAP when clear is called?"

No. Quite the opposite. The focus is on the "clear" method and to make that non-blocking. Once clear terminates the game start sequence can begin. Clear blocking until refresh completes is the thing we are changing.

Looking at it, the locking is as simple as it will be. The critical code path previously was just this variable assignment: `s_cachedGameModel = model;`. Method level synchronize blocks protected this code path. By taking those away, we now need a lock object to take its place. But also, to make the ordering work, we need a flag the clear method can set so that the clear method may be non-blocking. This flag is also shared memory and so also now on the critical code path. Thus we need a lock to procect the setting and checking of the flag, and also for the variable assignment.

"I think the changes have become overly complex. "

I obviously disagree. The prior locking solution  was over simplified and not without its defects. Synchronization and locking are complicated. Hence the reasons design patterns evolved that circumvent completely these locking issues, like avoid shared memory, static, and mutable objects.


"But can't calling clear 2 times in a row cause a NPE on cacheClearedMessenger?"

Calling clear twice would be a bug. Clear is called when a game starts, refresh when the application launches or a game closes. We launch one game per JVM, so the two calls not being sequenced properly would be a bug (and hence why there is so much synchronization code present). I'ved added the NPE check anyways, mainly for the benefit of a reader so they don't have to know that the double clear is impossible.

ron-murhammer commented:
@DanVanAtta Isn't canceling refresh so clear can proceed essentially accomplishing the same goal as making clear non-blocking so it doesn't wait on refresh? Am I missing something?

My understanding of your changes is that its doing 2 related but separate things:
1. Using cacheClearedMessenger to abort the refresh call this should make it so if refresh is in-progress that it completes quickly as soon as clear is called so the clear can then proceed
2. Making clear non-blocking so that it doesn't wait on refresh. If we remove the cacheClearedMessenger then this makes sense since refresh takes a while.

I don't think we need both of those changes or at least I'm failing to understand why.

ron-murhammer commented:
@DanVanAtta Here is an example of what I'm thinking should work and I briefly tested this as well:

```
  // any methods touching s_cachedGameModel should be both static and synchronized
  public synchronized static void refreshNewGameChooserModel() {
    clearNewGameChooserModel();
    cacheClearedMessenger = new ClearGameChooserCacheMessenger();
    s_cachedGameModel = new NewGameChooserModel(cacheClearedMessenger);
  }

  // any methods touching s_cachedGameModel should be both static and synchronized
  public static void clearNewGameChooserModel() {
    if (cacheClearedMessenger != null) {
      cacheClearedMessenger.sendCancel();
    }
    synchronizedClear();
  }

  private synchronized static void synchronizedClear() {
    if (s_cachedGameModel != null) {
      s_cachedGameModel.clear();
      s_cachedGameModel = null;
      cacheClearedMessenger = null;
    }
  }
```

DanVanAtta commented:
The more important change is to get rid of the blocking call. Without that the max wait delay for game launch is still seconds rather than nanoseconds. We've one map that is nearly a quarter gig, it could be a few seconds to parse that which in your example would still be a multi-second delay.

The cache clear messenger helps in the case when we are relaunching and helps avoid any build ups of background processing queue.  Consider the rapid restart scenario and picture what the un-cancelled background operations would be doing. That degenerate scenario is a good example to show some of the weaknesses in the current blocking design.

Ultimately the background operations, as long as they are not blocking, are only wasteful rather than UX impacting.  The resources *perhaps* could delay game launch since unzipping all map files will consume a lot of memory, memory not then available for the game launch. I did see that disk and CPU usage were a *lot* better after adding the cache clear messenger. Since we're so concerned about clearing out this variable (s_cachedGameModel) to save 5-10% of total memory usage, it seemed reasonable to add the extra small step to avoid the wasted processing as well.

ron-murhammer commented:
@DanVanAtta So I think we'll have to agree to disagree at this point. I'd prefer simpler code with just one of the 2 changes as I feel that gets us 90-95% of the benefit. Using the code I posted above makes the launch pretty much instant for me and worse case is as you posted above that it just started parsing a large map then you maybe wait a few seconds at most. That being said I think the code is better than what we have.

@veqryn This one is up to you now.

DanVanAtta commented:
Careful of benchmarks of map launch speed. It depends a lot on hard disk performance, you could be seeing a full 1000x times faster than other nominal cases. I replaced a hard drive that was 8 years old 3 weeks ago with a SSD that is a full 1000x times faster. Map parsing dropped from 5.3 to <0.1 seconds. Bottom line, slower hard drives are out there and the times are *much* slower for them.

DanVanAtta commented:
I found a way to cap how long it takes to parse any one map, @ron-murhammer  your suggestion then makes sense and I added a commit taking it. I also updated the PR title to reflect that clear will be blocking.

FYI: commits  b75bc44 and db78c05 are the same, just a cleaned up commit message.

ron-murhammer commented:
@DanVanAtta Nice work! Looks like the changes line up very closely now with what I was thinking.

ron-murhammer commented:
@veqryn I'm going to merge this now as I've tested it and read through it pretty thoroughly. Feel free to look over it as well when you find the time.
